libraries/common/array.effekt

```diff
@@ -1,90 +1,120 @@
-  ml "Array.array (${size}, ${default})"
-  default {
-    val arr = emptyArray[T](size)
-    each(0, size) { i =>
-      put(arr, i, default)
-    }
-    return arr;
-  }
-
-/**
- * Builds a new Array of size `size` from a computation `index` which gets an index
- * and returns a value that will be on that position in the resulting array
- */
-def build[T](size: Int) { index: Int => T }: Array[T] = {
-  val arr = emptyArray[T](size)
-  each(0, size) { i =>
-    put(arr, i, index(i))
-  }
-  return arr
-}
+extern type Array[T]
 
 /**
- * Converts a List `list` to an Array
+ * Violated Postcondition: Fields contain values
  */
-def fromList[T](list: List[T]): Array[T] = {
-  val listSize = list.size();
-  val arr = emptyArray(listSize);
-
-  foreachIndex(list) { (i, head) =>
-    put(arr, i, head)
-  }
-  return arr;
-}
-
-extern pure def emptyArray[T](size: Int): Array[T] =
+extern global def unsafeFresh[T](size: Int): Array[T] =
   js "(new Array(${size}))"
   chez "(make-vector ${size})" // creates an array filled with 0s on CS
-
-def get[T](arr: Array[T], index: Int): Option[T] =
-  if (index < 0 || index >= arr.size) None()
-  else undefinedToOption(arr.unsafeGet(index))
+  llvm """
+    %z = call %Pos @c_array_new(%Int %size)
+    ret %Pos %z
+  """
 
 extern pure def size[T](arr: Array[T]): Int =
   js "${arr}.length"
   chez "(vector-length ${arr})"
   ml "Array.length ${arr}"
+  llvm """
+    %z = call %Int @c_array_size(%Pos %arr)
+    ret %Int %z
+  """
 
-extern io def unsafeGet[T](arr: Array[T], index: Int): T =
+/**
+ * Unchecked Precondition: Index in bounds
+ */
+extern global def unsafeGet[T](arr: Array[T], index: Int): T =
   js "${arr}[${index}]"
   chez "(vector-ref ${arr} ${index})"
-  ml "SOME (Array.sub (${arr}, ${index}))"
+  ml "Array.sub (${arr}, ${index})"
+  llvm """
+    %z = call %Pos @c_array_get(%Pos %arr, %Int %index)
+    ret %Pos %z
+  """
 
 extern js """
-function put(arr, index, value) {
+function array$set(arr, index, value) {
   arr[index] = value;
   return $effekt.unit
 }
 """
 
-def put[T](arr: Array[T], index: Int, value: T): Unit =
-  if (index >= 0 && index < arr.size) unsafePut(arr, index, value)
-
-extern io def unsafePut[T](arr: Array[T], index: Int, value: T): Unit =
-  js "put(${arr}, ${index}, ${value})"
+/**
+ * Unchecked Precondition: Index in bounds
+ */
+extern global def unsafeSet[T](arr: Array[T], index: Int, value: T): Unit =
+  js "array$set(${arr}, ${index}, ${value})"
   chez "(begin (vector-set! ${arr} ${index} ${value}) #f)"
   ml "Array.update (${arr}, ${index}, ${value})"
+  llvm """
+    %z = call %Pos @c_array_set(%Pos %arr, %Int %index, %Pos %value)
+    ret %Pos %z
+  """
 
-extern pure def copy[T](arr: Array[T]): Array[T] =
+extern global def copy[T](arr: Array[T]): Array[T] =
   js "[...${arr}]"
 
+// Derived operations:
+
+/**
+ * Creates a new Array of size `size` filled with the value `init`
+ */
+extern global def fresh[T](size: Int, init: T): Array[T] =
```

Comment:
Note that `fill` is also already used for the same operation on lists — it might be worth it to think about synchronizing the two names.

Reply:
Hmm.. here we used `fresh` since it is the same name as for `Ref`.